File: composeApp/src/commonMain/kotlin/com/hhp227/concafe/presentation/main/cafemanagement/cafeinfo/CafeInfoEditViewModel.kt
package com.hhp227.concafe.presentation.main.cafemanagement.cafeinfo

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import com.hhp227.concafe.domain.common.AppResult
import com.hhp227.concafe.domain.model.CafeRegistrationDraft
import com.hhp227.concafe.domain.model.CafeInfoUpdate
import com.hhp227.concafe.domain.model.ConceptType
import com.hhp227.concafe.domain.model.GeoPoint
import com.hhp227.concafe.domain.model.Region
import com.hhp227.concafe.core.util.TimeUtils
import com.hhp227.concafe.domain.usecase.CreateCafeRegistrationClaimUseCase
import com.hhp227.concafe.domain.usecase.GetCafeDetailUseCase
import com.hhp227.concafe.domain.usecase.UploadImageUseCase
import com.hhp227.concafe.domain.usecase.UpdateCafeInfoUseCase

class CafeInfoEditViewModel(
    private val cafeId: String?,
    private val isRegistrationMode: Boolean,
    private val createCafeRegistrationClaimUseCase: CreateCafeRegistrationClaimUseCase,
    private val getCafeDetailUseCase: GetCafeDetailUseCase,
    private val updateCafeInfoUseCase: UpdateCafeInfoUseCase,
    private val uploadImageUseCase: UploadImageUseCase
) : ViewModel() {
    private val _uiState = MutableStateFlow(CafeInfoEditUiState(isRegistrationMode = isRegistrationMode))
    val uiState = _uiState.asStateFlow()

    private val _event = MutableSharedFlow<CafeInfoEvent>(replay = 0)
    val event = _event.asSharedFlow()

    private fun loadCafeInfo() {
        val targetCafeId = cafeId ?: return
        _uiState.update {
            it.copy(
                isLoading = true,
                infoMessage = null
            )
        }
        viewModelScope.launch {
            when (val result = getCafeDetailUseCase.invoke(targetCafeId)) {
                is AppResult.Success -> {
                    val detail = result.data.detail
                    val parsedHours = parseBusinessHours(detail.businessHours)
                    val cafeImages = detail.images.filter { image -> image.isNotBlank() }
                    _uiState.update {
                        it.copy(
                            detail = detail,
                            isLoading = false,
                            cafeName = detail.cafe.name,
                            cafeDescription = detail.cafe.desc,
                            conceptType = normalizeConceptType(detail.cafe.conceptType),
                            representativeImageUrl = cafeImages.firstOrNull() ?: detail.cafe.thumbnailImage,
                            galleryImages = cafeImages.drop(1).take(_uiState.value.galleryMaxCount),
                            address = detail.cafe.region.address,
                            mapLatitude = detail.cafe.region.location.latitude,
                            mapLongitude = detail.cafe.region.location.longitude,
                            contactNumber = detail.phoneNumber
                                .takeUnless { phone -> phone == CONTACT_PLACEHOLDER }
                                .orEmpty(),
                            weekdayOpen = parsedHours.weekdayOpen,
                            weekdayClose = parsedHours.weekdayClose,
                            weekendOpen = parsedHours.weekendOpen,
                            weekendClose = parsedHours.weekendClose,
                            infoMessage = null
                        )
                    }
                }
                is AppResult.Failure -> {
                    _uiState.update {
                        it.copy(
                            detail = null,
                            isLoading = false,
                            infoMessage = MSG_LOAD_FAILED
                        )
                    }
                }
            }
        }
    }

    private fun saveCafeInfo() {
        val currentState = _uiState.value
        if (currentState.representativeImageUrl.isNullOrBlank() && currentState.galleryImages.none { it.isNotBlank() }) {
            _uiState.update {
                it.copy(
                    isImageRequiredAlertVisible = true,
                    infoMessage = MSG_IMAGE_REQUIRED_ONE_OR_MORE
                )
            }
            return
        }

        if (isRegistrationMode) {
            submitCafeRegistration()
            return
        }

        val targetCafeId = cafeId ?: return
        _uiState.update { it.copy(isSaving = true, infoMessage = null) }

        viewModelScope.launch {
            val uploadedRepresentativeImage = if (currentState.representativeImageUrl.isNullOrBlank()) {
                null
            } else {
                uploadImage(currentState.representativeImageUrl, "cafes/representative") ?: return@launch
            }
            val uploadedGalleryImages = buildList {
                for (image in currentState.galleryImages.filter { it.isNotBlank() }) {
                    val uploaded = uploadImage(image, "cafes/gallery") ?: return@launch
                    add(uploaded)
                }
            }
            when (
                val result = updateCafeInfoUseCase.invoke(
                    CafeInfoUpdate(
                        cafeId = targetCafeId,
                        name = currentState.cafeName,
                        description = currentState.cafeDescription,
                        conceptType = normalizeConceptType(currentState.conceptType),
                        representativeImageUrl = uploadedRepresentativeImage,
                        galleryImages = uploadedGalleryImages,
                        location = GeoPoint(
                            latitude = currentState.mapLatitude,
                            longitude = currentState.mapLongitude
                        ),
                        address = currentState.address,
                        contactNumber = currentState.contactNumber,
                        weekdayOpen = currentState.weekdayOpen,
                        weekdayClose = currentState.weekdayClose,
                        weekendOpen = currentState.weekendOpen,
                        weekendClose = currentState.weekendClose
                    )
                )
            ) {
                is AppResult.Success -> {
                    val detail = result.data
                    val parsedHours = parseBusinessHours(detail.businessHours)
                    val cafeImages = detail.images.filter { image -> image.isNotBlank() }
                    _uiState.update {
                        it.copy(
                            detail = detail,
                            isSaving = false,
                            cafeName = detail.cafe.name,
                            cafeDescription = detail.cafe.desc,
                            conceptType = normalizeConceptType(detail.cafe.conceptType),
                            representativeImageUrl = cafeImages.firstOrNull() ?: detail.cafe.thumbnailImage,
                            galleryImages = cafeImages.drop(1).take(_uiState.value.galleryMaxCount),
                            address = detail.cafe.region.address,
                            mapLatitude = detail.cafe.region.location.latitude,
                            mapLongitude = detail.cafe.region.location.longitude,
                            contactNumber = detail.phoneNumber
                                .takeUnless { phone -> phone == CONTACT_PLACEHOLDER }
                                .orEmpty(),
                            weekdayOpen = parsedHours.weekdayOpen,
                            weekdayClose = parsedHours.weekdayClose,
                            weekendOpen = parsedHours.weekendOpen,
                            weekendClose = parsedHours.weekendClose,
                            infoMessage = null
                        )
                    }
                    _event.emit(CafeInfoEvent.NavigateBack)
                }
                is AppResult.Failure -> {
                    _uiState.update {
                        it.copy(
                            isSaving = false,
                            infoMessage = MSG_SAVE_FAILED
                        )
                    }
                }
            }
        }
    }

    private fun submitCafeRegistration() {
        val currentState = _uiState.value
        if (currentState.representativeImageUrl.isNullOrBlank()) {
            _uiState.update {
                it.copy(
                    isImageRequiredAlertVisible = true,
                    infoMessage = MSG_REGISTRATION_REP_REQUIRED
                )
            }
            return
        }
        _uiState.update { it.copy(isSaving = true, infoMessage = null) }

        viewModelScope.launch {
            val uploadedRepresentativeImage = uploadImage(currentState.representativeImageUrl, "cafes/representative")
                ?: return@launch
            when (
                val result = createCafeRegistrationClaimUseCase.invoke(
                    CafeRegistrationDraft(
                        name = currentState.cafeName.trim(),
                        description = currentState.cafeDescription.trim(),
                        region = Region(
                            country = "KR",
                            city = "Seoul",
                            address = currentState.address.trim(),
                            location = GeoPoint(
                                latitude = currentState.mapLatitude,
                                longitude = currentState.mapLongitude
                            )
                        ),
                        thumbnailImage = uploadedRepresentativeImage,
                        conceptType = normalizeConceptType(currentState.conceptType),
                        businessHours = formatBusinessHours(currentState),
                        phoneNumber = currentState.contactNumber.trim()
                    )
                )
            ) {
                is AppResult.Success -> {
                    _uiState.update { it.copy(isSaving = false, infoMessage = null) }
                    _event.emit(CafeInfoEvent.NavigateBack)
                }
                is AppResult.Failure -> {
                    _uiState.update {
                        it.copy(
                            isSaving = false,
                            infoMessage = result.error.toString()
                        )
                    }
                }
            }
        }
    }

    private fun formatBusinessHours(state: CafeInfoEditUiState): String {
        val weekday = listOf(state.weekdayOpen, state.weekdayClose).all { it.isNotBlank() }
        val weekend = listOf(state.weekendOpen, state.weekendClose).all { it.isNotBlank() }
        return when {
            weekday && weekend && state.weekdayOpen == state.weekendOpen && state.weekdayClose == state.weekendClose ->
                "매일 ${state.weekdayOpen} - ${state.weekdayClose}"
            weekday && weekend ->
                "평일 ${state.weekdayOpen} - ${state.weekdayClose} / 주말 ${state.weekendOpen} - ${state.weekendClose}"
            weekday ->
                "평일 ${state.weekdayOpen} - ${state.weekdayClose}"
            weekend ->
                "주말 ${state.weekendOpen} - ${state.weekendClose}"
            else -> ""
        }
    }

    fun onAction(action: CafeInfoEditAction) {
        when (action) {
            CafeInfoEditAction.ClickBack -> {
                viewModelScope.launch { _event.emit(CafeInfoEvent.NavigateBack) }
            }
            is CafeInfoEditAction.ChangeCafeName -> _uiState.update { it.copy(cafeName = action.value) }
            is CafeInfoEditAction.ChangeCafeDescription -> _uiState.update { it.copy(cafeDescription = action.value) }
            is CafeInfoEditAction.ChangeConceptType -> _uiState.update { it.copy(conceptType = normalizeConceptType(action.value)) }
            is CafeInfoEditAction.ChangeAddress -> _uiState.update { it.copy(address = action.value) }
            is CafeInfoEditAction.SetPinnedLocation -> _uiState.update {
                it.copy(
                    mapLatitude = action.latitude,
                    mapLongitude = action.longitude
                )
            }
            is CafeInfoEditAction.ChangeContactNumber -> _uiState.update { it.copy(contactNumber = action.value) }
            is CafeInfoEditAction.ChangeWeekdayOpen -> _uiState.update { it.copy(weekdayOpen = action.value) }
            is CafeInfoEditAction.ChangeWeekdayClose -> _uiState.update { it.copy(weekdayClose = action.value) }
            is CafeInfoEditAction.ChangeWeekendOpen -> _uiState.update { it.copy(weekendOpen = action.value) }
            is CafeInfoEditAction.ChangeWeekendClose -> _uiState.update { it.copy(weekendClose = action.value) }
            is CafeInfoEditAction.SelectRepresentativeImage -> {
                _uiState.update { it.copy(representativeImageUrl = action.imageUrl, isImageRequiredAlertVisible = false) }
            }
            is CafeInfoEditAction.AddGalleryImage -> {
                if (action.imageUrl.isBlank()) {
                    return
                }
                val maxCount = _uiState.value.galleryMaxCount
                if (_uiState.value.galleryImages.size >= maxCount) {
                    showInfo("$MSG_GALLERY_MAX_EXCEEDED:$maxCount")
                    return
                }
                _uiState.update { state ->
                    state.copy(
                        galleryImages = state.galleryImages + action.imageUrl,
                        isImageRequiredAlertVisible = false
                    )
                }
            }
            is CafeInfoEditAction.RemoveGalleryImage -> {
                _uiState.update { state ->
                    val updated = state.galleryImages.toMutableList().also { it.removeAt(action.index) }

                    state.copy(galleryImages = updated)
                }
            }
            CafeInfoEditAction.ClickRepresentativeImage -> showInfo(MSG_REP_UPLOAD_NEXT_STEP)
            CafeInfoEditAction.ClickAddGalleryImage -> showInfo(MSG_GALLERY_ADD_NEXT_STEP)
            CafeInfoEditAction.ClickPinLocation -> showInfo(MSG_PIN_LOCATION_HINT)
            CafeInfoEditAction.ClickManageExceptionDates -> showInfo(MSG_EXCEPTION_NEXT_STEP)
            CafeInfoEditAction.DismissImageRequiredAlert -> _uiState.update { it.copy(isImageRequiredAlertVisible = false) }
            CafeInfoEditAction.ClickSave -> saveCafeInfo()
            CafeInfoEditAction.DismissInfoMessage -> _uiState.update { it.copy(infoMessage = null) }
        }
    }

    private suspend fun uploadImage(imageUrl: String?, folder: String): String? {
        if (imageUrl.isNullOrBlank()) return null
        return when (val result = uploadImageUseCase.invoke(imageUrl, folder)) {
            is AppResult.Success -> result.data
            is AppResult.Failure -> {
                _uiState.update {
                    it.copy(
                        isSaving = false,
                        infoMessage = MSG_IMAGE_UPLOAD_FAILED
                    )
                }
                null
            }
        }
    }

    private fun showInfo(message: String) {
        _uiState.update { it.copy(infoMessage = message) }
    }

    private fun parseBusinessHours(businessHours: String): ParsedBusinessHours {
        val times = TimeUtils.extractNormalizedHourMinuteList(businessHours)
        val open = times.getOrNull(0).orEmpty()
        val close = times.getOrNull(1).orEmpty()
        val weekendOpen = times.getOrNull(2) ?: open
        val weekendClose = times.getOrNull(3) ?: close
        return ParsedBusinessHours(
            weekdayOpen = open,
            weekdayClose = close,
            weekendOpen = weekendOpen,
            weekendClose = weekendClose
        )
    }

    private data class ParsedBusinessHours(
        val weekdayOpen: String,
        val weekdayClose: String,
        val weekendOpen: String,
        val weekendClose: String
    )

    init {
        if (isRegistrationMode) {
            _uiState.update { it.copy(isLoading = false, infoMessage = null) }
        } else {
            loadCafeInfo()
        }
    }

    companion object {
        private const val CONTACT_PLACEHOLDER = "연락처 정보 준비중"
        private const val MSG_LOAD_FAILED = "cafeinfo_info_load_failed"
        private const val MSG_IMAGE_REQUIRED_ONE_OR_MORE = "cafeinfo_info_image_required_one_or_more"
        private const val MSG_SAVE_FAILED = "cafeinfo_info_save_failed"
        private const val MSG_REGISTRATION_REP_REQUIRED = "cafeinfo_info_registration_rep_required"
        private const val MSG_GALLERY_MAX_EXCEEDED = "cafeinfo_info_gallery_max_exceeded"
        private const val MSG_REP_UPLOAD_NEXT_STEP = "cafeinfo_info_rep_upload_next_step"
        private const val MSG_GALLERY_ADD_NEXT_STEP = "cafeinfo_info_gallery_add_next_step"
        private const val MSG_PIN_LOCATION_HINT = "cafeinfo_info_pin_location_hint"
        private const val MSG_EXCEPTION_NEXT_STEP = "cafeinfo_info_exception_next_step"
        private const val MSG_IMAGE_UPLOAD_FAILED = "cafeinfo_info_image_upload_failed"

        private fun normalizeConceptType(value: String): String {
            return ConceptType.fromRaw(value)?.name ?: ConceptType.MAID.name
        }
    }
}

File: iosApp/iosApp/Presentation/Main/CafeManagement/CafeInfoEdit/CafeInfoEditViewModel.swift
//
//  CafeInfoEditViewModel.swift
//  ConCafe
//
//  Created by 홍희표 on 2026/03/11.
//

import Foundation
import Combine
import Shared

@MainActor
final class CafeInfoEditViewModel: ObservableObject {
    private let cafeId: String?

    private let isRegistrationMode: Bool

    private let getCafeDetailUseCase: GetCafeDetailUseCase

    private let createCafeRegistrationClaimUseCase: CreateCafeRegistrationClaimUseCase

    private let updateCafeInfoUseCase: UpdateCafeInfoUseCase

    private let uploadImageUseCase: UploadImageUseCase

    @Published private(set) var uiState = CafeInfoEditUiState()

    let event = PassthroughSubject<CafeInfoEditEvent, Never>()

    private var loadTask: Task<Void, Never>?

    private func loadCafeInfo() {
        guard let cafeId else { return }
        loadTask?.cancel()
        uiState.isLoading = true
        uiState.infoMessage = nil

        loadTask = Task {
            do {
                let result = try await getCafeDetailUseCase.invoke(cafeId: cafeId)

                if let success = result as? AppResultSuccess<AnyObject>,
                   let feed = success.data as? CafeDetailFeed {
                    let detail = feed.detail
                    let parsedHours = parseBusinessHours(detail.businessHours)
                    let cafeImages = detail.images.filter { !$0.isEmpty }

                    uiState.detail = detail
                    uiState.isLoading = false
                    uiState.cafeName = detail.cafe.name
                    uiState.cafeDescription = detail.cafe.desc
                    uiState.representativeImageUrl = cafeImages.first ?? detail.cafe.thumbnailImage
                    uiState.galleryImages = Array(cafeImages.dropFirst().prefix(uiState.galleryMaxCount))
                    uiState.address = detail.cafe.region.address
                    uiState.mapLatitude = detail.cafe.region.location.latitude
                    uiState.mapLongitude = detail.cafe.region.location.longitude
                    uiState.contactNumber = detail.phoneNumber == MessageKey.contactPlaceholder ? "" : detail.phoneNumber
                    uiState.weekdayOpen = parsedHours.weekdayOpen
                    uiState.weekdayClose = parsedHours.weekdayClose
                    uiState.weekendOpen = parsedHours.weekendOpen
                    uiState.weekendClose = parsedHours.weekendClose
                    uiState.infoMessage = nil
                } else {
                    uiState.detail = nil
                    uiState.isLoading = false
                    uiState.infoMessage = MessageKey.loadFailed
                }
            } catch {
                if Task.isCancelled { return }
                uiState.detail = nil
                uiState.isLoading = false
                uiState.infoMessage = MessageKey.loadFailed
            }
        }
    }

    private func saveCafeInfo() {
        if (uiState.representativeImageUrl?.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty ?? true)
            && uiState.galleryImages.allSatisfy({ $0.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty }) {
            uiState.isImageRequiredAlertVisible = true
            uiState.infoMessage = MessageKey.imageRequiredOneOrMore
            return
        }

        if isRegistrationMode {
            submitCafeRegistration()
            return
        }

        guard let cafeId else { return }
        uiState.isSaving = true
        uiState.infoMessage = nil

        loadTask?.cancel()
        loadTask = Task {
            do {
                let uploadedRepresentativeImage = try await uploadImageIfNeeded(
                    uiState.representativeImageUrl,
                    folder: "cafes/representative"
                )
                let uploadedGalleryImages = try await uploadImagesIfNeeded(
                    uiState.galleryImages,
                    folder: "cafes/gallery"
                )
                let result = try await updateCafeInfoUseCase.invoke(
                    update: CafeInfoUpdate(
                        cafeId: cafeId,
                        name: uiState.cafeName,
                        description: uiState.cafeDescription,
                        representativeImageUrl: uploadedRepresentativeImage,
                        galleryImages: uploadedGalleryImages,
                        location: GeoPoint(
                            latitude: uiState.mapLatitude,
                            longitude: uiState.mapLongitude
                        ),
                        address: uiState.address,
                        contactNumber: uiState.contactNumber,
                        weekdayOpen: uiState.weekdayOpen,
                        weekdayClose: uiState.weekdayClose,
                        weekendOpen: uiState.weekendOpen,
                        weekendClose: uiState.weekendClose
                    )
                )

                if let success = result as? AppResultSuccess<AnyObject>,
                   let detail = success.data as? CafeDetail {
                    let parsedHours = parseBusinessHours(detail.businessHours)
                    let cafeImages = detail.images.filter { !$0.isEmpty }

                    uiState.detail = detail
                    uiState.isSaving = false
                    uiState.cafeName = detail.cafe.name
                    uiState.cafeDescription = detail.cafe.desc
                    uiState.representativeImageUrl = cafeImages.first ?? detail.cafe.thumbnailImage
                    uiState.galleryImages = Array(cafeImages.dropFirst().prefix(uiState.galleryMaxCount))
                    uiState.address = detail.cafe.region.address
                    uiState.mapLatitude = detail.cafe.region.location.latitude
                    uiState.mapLongitude = detail.cafe.region.location.longitude
                    uiState.contactNumber = detail.phoneNumber == MessageKey.contactPlaceholder ? "" : detail.phoneNumber
                    uiState.weekdayOpen = parsedHours.weekdayOpen
                    uiState.weekdayClose = parsedHours.weekdayClose
                    uiState.weekendOpen = parsedHours.weekendOpen
                    uiState.weekendClose = parsedHours.weekendClose
                    uiState.infoMessage = nil
                    event.send(.navigateBack)
                } else {
                    uiState.isSaving = false
                    uiState.infoMessage = MessageKey.saveFailed
                }
            } catch {
                if Task.isCancelled { return }
                uiState.isSaving = false
                uiState.infoMessage = MessageKey.saveFailed
            }
        }
    }

    private func submitCafeRegistration() {
        if uiState.representativeImageUrl?.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty ?? true {
            uiState.isImageRequiredAlertVisible = true
            uiState.infoMessage = MessageKey.registrationRepRequired
            return
        }

        uiState.isSaving = true
        uiState.infoMessage = nil

        loadTask?.cancel()
        loadTask = Task {
            do {
                let uploadedRepresentativeImage = try await uploadImageIfNeeded(
                    uiState.representativeImageUrl,
                    folder: "cafes/representative"
                )
                let result = try await createCafeRegistrationClaimUseCase.invoke(
                    draft: CafeRegistrationDraft(
                        name: uiState.cafeName.trimmingCharacters(in: .whitespacesAndNewlines),
                        description: uiState.cafeDescription.trimmingCharacters(in: .whitespacesAndNewlines),
                        region: Region(
                            country: "KR",
                            city: "Seoul",
                            address: uiState.address.trimmingCharacters(in: .whitespacesAndNewlines),
                            location: GeoPoint(
                                latitude: uiState.mapLatitude,
                                longitude: uiState.mapLongitude
                            )
                        ),
                        thumbnailImage: uploadedRepresentativeImage,
                        conceptType: "MAID",
                        businessHours: formatBusinessHours(),
                        phoneNumber: uiState.contactNumber.trimmingCharacters(in: .whitespacesAndNewlines)
                    )
                )

                if result is AppResultSuccess<AnyObject> {
                    uiState.isSaving = false
                    uiState.infoMessage = nil
                    event.send(.navigateBack)
                } else if let failure = result as? AppResultFailure {
                    uiState.isSaving = false
                    uiState.infoMessage = "\(failure.error)"
                }
            } catch {
                if Task.isCancelled { return }
                uiState.isSaving = false
                uiState.infoMessage = error.localizedDescription
            }
        }
    }

    private func formatBusinessHours() -> String {
        let weekday = !uiState.weekdayOpen.isEmpty && !uiState.weekdayClose.isEmpty
        let weekend = !uiState.weekendOpen.isEmpty && !uiState.weekendClose.isEmpty

        if weekday && weekend && uiState.weekdayOpen == uiState.weekendOpen && uiState.weekdayClose == uiState.weekendClose {
            return "매일 \(uiState.weekdayOpen) - \(uiState.weekdayClose)"
        }
        if weekday && weekend {
            return "평일 \(uiState.weekdayOpen) - \(uiState.weekdayClose) / 주말 \(uiState.weekendOpen) - \(uiState.weekendClose)"
        }
        if weekday {
            return "평일 \(uiState.weekdayOpen) - \(uiState.weekdayClose)"
        }
        if weekend {
            return "주말 \(uiState.weekendOpen) - \(uiState.weekendClose)"
        }
        return ""
    }

    func onAction(_ action: CafeInfoEditAction) {
        switch action {
        case .clickBack:
            event.send(.navigateBack)
        case .changeCafeName(let value):
            uiState.cafeName = value
        case .changeCafeDescription(let value):
            uiState.cafeDescription = value
        case .changeAddress(let value):
            uiState.address = value
        case .changeContactNumber(let value):
            uiState.contactNumber = value
        case .changeWeekdayOpen(let value):
            uiState.weekdayOpen = value
        case .changeWeekdayClose(let value):
            uiState.weekdayClose = value
        case .changeWeekendOpen(let value):
            uiState.weekendOpen = value
        case .changeWeekendClose(let value):
            uiState.weekendClose = value
        case .selectRepresentativeImage(let imageUrl):
            uiState.representativeImageUrl = imageUrl
        case .addGalleryImage(let imageUrl):
            if uiState.galleryImages.count >= uiState.galleryMaxCount {
                showInfo("\(MessageKey.galleryMaxExceeded):\(uiState.galleryMaxCount)")
                return
            }
            if imageUrl.isEmpty { return }
            uiState.galleryImages.append(imageUrl)
        case .clickRepresentativeImage:
            break
        case .clickAddGalleryImage:
            break
        case .clickPinLocation:
            showInfo(MessageKey.pinLocationHint)
        case .setPinnedLocation(let latitude, let longitude):
            uiState.mapLatitude = latitude
            uiState.mapLongitude = longitude
        case .clickManageExceptionDates:
            showInfo(MessageKey.exceptionNextStep)
        case .dismissImageRequiredAlert:
            uiState.isImageRequiredAlertVisible = false
        case .clickSave:
            saveCafeInfo()
        case .dismissInfoMessage:
            uiState.infoMessage = nil
        }
    }

    private func showInfo(_ message: String) {
        uiState.infoMessage = message
    }

    private func parseBusinessHours(_ businessHours: String) -> (weekdayOpen: String, weekdayClose: String, weekendOpen: String, weekendClose: String) {
        let times = TimeUtils.extractNormalizedHourMinuteList(from: businessHours)
        let weekdayOpen = times.indices.contains(0) ? times[0] : ""
        let weekdayClose = times.indices.contains(1) ? times[1] : ""
        let weekendOpen = times.indices.contains(2) ? times[2] : weekdayOpen
        let weekendClose = times.indices.contains(3) ? times[3] : weekdayClose
        return (weekdayOpen, weekdayClose, weekendOpen, weekendClose)
    }

    init(
        cafeId: String? = nil,
        isRegistrationMode: Bool = false,
        createCafeRegistrationClaimUseCase: CreateCafeRegistrationClaimUseCase = KoinInitializerKt.resolveCreateCafeRegistrationClaimUseCase(),
        getCafeDetailUseCase: GetCafeDetailUseCase = KoinInitializerKt.resolveGetCafeDetailUseCase(),
        updateCafeInfoUseCase: UpdateCafeInfoUseCase = KoinInitializerKt.resolveUpdateCafeInfoUseCase(),
        uploadImageUseCase: UploadImageUseCase = KoinInitializerKt.resolveUploadImageUseCase()
    ) {
        self.cafeId = cafeId
        self.isRegistrationMode = isRegistrationMode
        self.createCafeRegistrationClaimUseCase = createCafeRegistrationClaimUseCase
        self.getCafeDetailUseCase = getCafeDetailUseCase
        self.updateCafeInfoUseCase = updateCafeInfoUseCase
        self.uploadImageUseCase = uploadImageUseCase
        self.uiState.isRegistrationMode = isRegistrationMode

        if isRegistrationMode {
            uiState.isLoading = false
        } else {
            loadCafeInfo()
        }
    }

    deinit {
        loadTask?.cancel()
    }

    private func uploadImageIfNeeded(_ imageUrl: String?, folder: String) async throws -> String? {
        guard let imageUrl, !imageUrl.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            return nil
        }
        let result = try await uploadImageUseCase.invoke(localPath: imageUrl, folder: folder)
        if let success = result as? AppResultSuccess<AnyObject>, let data = success.data as? String {
            return data
        }
        if let failure = result as? AppResultFailure, let validation = failure.error as? AppErrorValidationFailed {
            throw NSError(domain: "CafeInfoEdit", code: 1, userInfo: [NSLocalizedDescriptionKey: validation.reason])
        }
        throw NSError(domain: "CafeInfoEdit", code: 1, userInfo: [NSLocalizedDescriptionKey: MessageKey.imageUploadFailed])
    }

    private func uploadImagesIfNeeded(_ imageUrls: [String], folder: String) async throws -> [String] {
        var uploaded: [String] = []
        for imageUrl in imageUrls where !imageUrl.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            if let result = try await uploadImageIfNeeded(imageUrl, folder: folder) {
                uploaded.append(result)
            }
        }
        return uploaded
    }

    private enum MessageKey {
        static let loadFailed = "cafeinfo_info_load_failed"
        static let imageRequiredOneOrMore = "cafeinfo_info_image_required_one_or_more"
        static let saveFailed = "cafeinfo_info_save_failed"
        static let registrationRepRequired = "cafeinfo_info_registration_rep_required"
        static let galleryMaxExceeded = "cafeinfo_info_gallery_max_exceeded"
        static let pinLocationHint = "cafeinfo_info_pin_location_hint"
        static let exceptionNextStep = "cafeinfo_info_exception_next_step"
        static let imageUploadFailed = "cafeinfo_info_image_upload_failed"
        static let contactPlaceholder = "연락처 정보 준비중"
    }
}
